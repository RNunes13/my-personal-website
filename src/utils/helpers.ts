/**
 * Get a RGB color from the hexadecimal.
 * @param {string} hex Hex value in the format `#181923`.
 * @public
 */
export const getRgbFromHex = (hex = '') => {
  const hexCode = hex.replace('#', '')

  if (hexCode.length !== 6) {
    throw Error('Only six-digit hex colors are allowed.')
  }

  const rgbHex = hexCode.match(/.{1,2}/g) as RegExpMatchArray

  return [
    parseInt(rgbHex[0], 16),
    parseInt(rgbHex[1], 16),
    parseInt(rgbHex[2], 16),
  ]
}
