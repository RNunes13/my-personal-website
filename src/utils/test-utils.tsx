import { FC, ReactElement } from 'react'
import { ThemeProvider } from '@emotion/react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import { theme } from 'styles/theme/Theme'

interface IProviders {
  children: ReactElement
}

const Providers: FC<IProviders> = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const renderWithTheme = (component: ReactElement): ReactTestRenderer =>
  renderer.create(<Providers>{component}</Providers>)

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(ui, { wrapper: Providers, ...options })

const renderWithUser = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => ({
  user: userEvent.setup(),
  ...render(ui, { wrapper: Providers, ...options }),
})

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { customRender as render }
export { renderWithUser }
