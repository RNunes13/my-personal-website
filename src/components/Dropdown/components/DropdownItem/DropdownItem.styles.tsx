import styled from '@emotion/styled'
import { bodySmallBold } from 'styles/mixins'
import { newShade } from 'utils/helpers'
import { Icon as IconComp } from 'components'

export const Item = styled.li`
  display: flex;
  align-items: center;
  padding: 5px;
  gap: 8px;
  cursor: pointer;
  border-radius: ${({ theme }) => theme.radius};
  color: ${({ theme }) => theme.colors.base_0};
  background-color: ${({ theme }) => newShade(theme.colors.primary, 20)};
  transition: background-color 200ms ${({ theme }) => theme.transitions.decelerate};

  &:hover {
    background-color: ${({ theme }) => newShade(theme.colors.primary, 40)};
  }
`

export const Icon = styled(IconComp)`
  flex: 0 0 auto;
`

export const Label = styled.span`
  ${bodySmallBold()}
  flex: 1 1 auto;
`
