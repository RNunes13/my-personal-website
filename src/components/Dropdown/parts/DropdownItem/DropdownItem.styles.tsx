import styled from '@emotion/styled'
import { bodySmallBold } from 'styles/mixins'
import { newShade } from 'utils/helpers'
import { Icon as IconComp } from 'components'

export const Item = styled.li<{ disabled?: boolean }>`
  cursor: pointer;
  display: flex;
  align-items: center;
  padding: 5px;
  gap: 8px;
  border-radius: ${({ theme }) => theme.radius};
  color: ${({ theme }) => theme.colors.base_0};
  background-color: ${({ theme }) => newShade(theme.colors.primary, 20)};

  transition: background-color 200ms
    ${({ theme }) => theme.transitions.decelerate};

  ${({ theme, disabled }) =>
    disabled &&
    `
    cursor: default;
    color: ${newShade(theme.colors.base_60, 90)};
  `}

  &:hover {
    background-color: ${({ theme, disabled }) =>
      !disabled && newShade(theme.colors.primary, 40)};
  }
`

export const Icon = styled(IconComp)`
  flex: 0 0 auto;
`

export const Label = styled.span`
  ${bodySmallBold()}
  flex: 1 1 auto;
`
