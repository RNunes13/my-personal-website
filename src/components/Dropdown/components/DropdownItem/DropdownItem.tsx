import React from 'react'
import * as Styled from './DropdownItem.styles'

export type IconType = {
  icon: string
  color?: string
  height?: number
  width?: number
}

interface DropdownItemProps {
  label: string
  startIcon?: string | IconType
  endIcon?: string | IconType
  onClick: () => void
}

export const getIconProps = (iconProp: string | IconType) => {
  if (typeof iconProp === 'string') return {
    icon: iconProp,
    fill: 'base_0',
  }

  const { icon, color, height, width } = iconProp

  return { icon, height, width, fill: color ?? 'base_0' }
}

export const DropdownItem: React.FC<DropdownItemProps> = ({
  label,
  startIcon,
  endIcon,
  onClick,
}) => (
  <Styled.Item onClick={onClick}>
    {!!startIcon && <Styled.Icon {...getIconProps(startIcon)} />}
    <Styled.Label>{label}</Styled.Label>
    {!!endIcon && <Styled.Icon {...getIconProps(endIcon)} />}
  </Styled.Item>
)
