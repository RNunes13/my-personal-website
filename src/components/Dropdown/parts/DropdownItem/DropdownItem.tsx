import React from 'react'
import * as Styled from './DropdownItem.styles'

export type IconType = {
  icon: string
  color?: string
  height?: number
  width?: number
}

export interface DropdownItemProps {
  label: string
  className?: string
  disabled?: boolean
  startIcon?: string | IconType
  endIcon?: string | IconType
  onClick?: () => void
}

export const getIconProps = (iconProp: string | IconType) => {
  if (typeof iconProp === 'string') return { icon: iconProp, fill: 'base_0' }

  const { icon, color, height, width } = iconProp

  return { icon, height, width, fill: color ?? 'base_0' }
}

export const DropdownItem: React.FC<DropdownItemProps> = ({
  label,
  startIcon,
  endIcon,
  disabled,
  className,
  onClick,
}) => (
  <Styled.Item
    className={className}
    disabled={disabled}
    onClick={onClick}
    data-testid="dropdownItem"
  >
    {!!startIcon && <Styled.Icon {...getIconProps(startIcon)} />}
    <Styled.Label>{label}</Styled.Label>
    {!!endIcon && <Styled.Icon {...getIconProps(endIcon)} />}
  </Styled.Item>
)
