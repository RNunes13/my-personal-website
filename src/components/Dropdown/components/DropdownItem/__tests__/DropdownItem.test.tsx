import { Suspense } from 'react'
import { renderWithUser, renderWithTheme } from 'utils/test-utils'
import { DropdownItem } from '../DropdownItem'
import type { DropdownItemProps, IconType } from '../DropdownItem'

const defaultProps: DropdownItemProps = {
  label: 'Item label',
  onClick: jest.fn(),
}

const getComponent = (props: Partial<DropdownItemProps> = {}) => (
  <Suspense fallback={<div>loading...</div>}>
    <DropdownItem {...defaultProps} {...props} />
  </Suspense>
)

describe('DropdownItem unit test', () => {
  it('should call the provide click functon', async () => {
    const onClick = jest.fn()
    const { getByTestId, user } = renderWithUser(getComponent({ onClick }))
    const item = getByTestId('dropdownItem')

    expect(onClick).not.toHaveBeenCalled()

    await user.click(item)

    expect(onClick).toHaveBeenCalled()
  })

  it('should render with start icon as string', () => {
    const tree = renderWithTheme(
      getComponent({ startIcon: 'system:menu' }),
    ).toJSON()
    expect(tree).toMatchSnapshot('DropdownItem - Start icon as string')
  })

  it('should render with start icon as object', () => {
    const startIcon: IconType = {
      icon: 'system:mennu',
      color: 'accent',
      height: 32,
      width: 32,
    }

    const component = getComponent({ startIcon })

    const tree = renderWithTheme(component).toJSON()
    expect(tree).toMatchSnapshot('DropdownItem - Start icon as object')
  })

  it('should render with end icon as string', () => {
    const tree = renderWithTheme(
      getComponent({ endIcon: 'system:menu' }),
    ).toJSON()
    expect(tree).toMatchSnapshot('DropdownItem - End icon as string')
  })

  it('should render with end icon as object', () => {
    const endIcon: IconType = {
      icon: 'system:mennu',
      color: 'primary',
      height: 16,
      width: 16,
    }

    const component = getComponent({ endIcon })

    const tree = renderWithTheme(component).toJSON()
    expect(tree).toMatchSnapshot('DropdownItem - Ennd icon as object')
  })
})
