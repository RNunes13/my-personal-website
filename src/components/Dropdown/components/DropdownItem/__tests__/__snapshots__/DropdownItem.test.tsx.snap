// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownItem unit test should render with end icon as object: DropdownItem - Ennd icon as object 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px;
  gap: 8px;
  cursor: pointer;
  border-radius: 4px;
  color: #FFFFFF;
  background-color: #2c2d37;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
  transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
}

.emotion-0:hover {
  background-color: #40414b;
}

.emotion-2 {
  font-family: Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.emotion-4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

<li
  className="emotion-0 emotion-1"
  data-testid="dropdownItem"
  onClick={[MockFunction]}
>
  <span
    className="emotion-2 emotion-3"
  >
    Item label
  </span>
  <svg
    className="emotion-4 emotion-5"
    data-testid="icon_system_mennu"
    fill="primary"
    height={16}
    icon="system:mennu"
    width={16}
  />
</li>
`;

exports[`DropdownItem unit test should render with end icon as string: DropdownItem - End icon as string 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px;
  gap: 8px;
  cursor: pointer;
  border-radius: 4px;
  color: #FFFFFF;
  background-color: #2c2d37;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
  transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
}

.emotion-0:hover {
  background-color: #40414b;
}

.emotion-2 {
  font-family: Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.emotion-4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

<li
  className="emotion-0 emotion-1"
  data-testid="dropdownItem"
  onClick={[MockFunction]}
>
  <span
    className="emotion-2 emotion-3"
  >
    Item label
  </span>
  <svg
    className="emotion-4 emotion-5"
    data-testid="icon_system_menu"
    fill="base_0"
    icon="system:menu"
  />
</li>
`;

exports[`DropdownItem unit test should render with start icon as object: DropdownItem - Start icon as object 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px;
  gap: 8px;
  cursor: pointer;
  border-radius: 4px;
  color: #FFFFFF;
  background-color: #2c2d37;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
  transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
}

.emotion-0:hover {
  background-color: #40414b;
}

.emotion-2 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.emotion-4 {
  font-family: Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

<li
  className="emotion-0 emotion-1"
  data-testid="dropdownItem"
  onClick={[MockFunction]}
>
  <svg
    className="emotion-2 emotion-3"
    data-testid="icon_system_mennu"
    fill="accent"
    height={32}
    icon="system:mennu"
    width={32}
  />
  <span
    className="emotion-4 emotion-5"
  >
    Item label
  </span>
</li>
`;

exports[`DropdownItem unit test should render with start icon as string: DropdownItem - Start icon as string 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px;
  gap: 8px;
  cursor: pointer;
  border-radius: 4px;
  color: #FFFFFF;
  background-color: #2c2d37;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
  transition: background-color 200ms cubic-bezier(0.0, 0.0, 0.2, 1);
}

.emotion-0:hover {
  background-color: #40414b;
}

.emotion-2 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.emotion-4 {
  font-family: Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

<li
  className="emotion-0 emotion-1"
  data-testid="dropdownItem"
  onClick={[MockFunction]}
>
  <svg
    className="emotion-2 emotion-3"
    data-testid="icon_system_menu"
    fill="base_0"
    icon="system:menu"
  />
  <span
    className="emotion-4 emotion-5"
  >
    Item label
  </span>
</li>
`;
